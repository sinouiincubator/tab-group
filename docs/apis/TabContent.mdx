---
name: TabContent
route: /api-tab-content
menu: API
---

import { Playground, Props } from 'docz';
import TabContent from '../../src/components/tab-content/TabContent';

# TabContent

<Props of={TabContent} />
