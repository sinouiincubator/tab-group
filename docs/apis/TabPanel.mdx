---
name: TabPanel
route: /api-tab-panel
menu: API
---

import { Playground, Props } from 'docz';
import TabPanel from '../../src/components/tab-panel/TabPanel';

# TabPanel

<Props of={TabPanel} />
