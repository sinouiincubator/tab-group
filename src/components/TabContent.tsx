import React, { useMemo } from 'react';
import classNames from 'classnames';
import styled from 'styled-components';
import TabContentContext from './TabContentContext';

export interface Props {
  children?: React.ReactNode;
  className?: string;
  active: number;
}

const Container = styled.div`
  overflow-x: hidden;
`;

const TabContentWrapper = styled.div`
  display: flex;
  transition: transform 0.35s cubic-bezier(0.15, 0.3, 0.25, 1) 0s;
`;

export default function TabContent(props: Props) {
  const { className, active, children } = props;
  const context = useMemo(() => ({ active }), [active]);
  const transform = useMemo(() => `translate3d(-${100 * active}%, 0px, 0px)`, [
    active,
  ]);

  return (
    <TabContentContext.Provider value={context}>
      <Container>
        <TabContentWrapper
          className={classNames('sinoui-tab-content', className)}
          style={{ transform }}
        >
          {children}
        </TabContentWrapper>
      </Container>
    </TabContentContext.Provider>
  );
}
