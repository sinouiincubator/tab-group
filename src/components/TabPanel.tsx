import React, { useContext, useMemo, useState, useEffect } from 'react';
import classNames from 'classnames';
import styled from 'styled-components';
import TabContentContext from './TabContentContext';

interface TabPanelProps {
  children: React.ReactNode;
  id: number | string;
  className?: string;
  style?: React.CSSProperties;
}

const TabPanelWrapper = styled.div`
  width: 100%;
  flex-shrink: 0;
`;

export default function TabPanel(props: TabPanelProps) {
  const { children, id, className, ...other } = props;
  const { active } = useContext(TabContentContext);
  const [isRendered, setIsRendered] = useState(false);
  const isShow = useMemo(() => active === id, [active, id]);

  useEffect(() => {
    if (isShow) {
      setIsRendered(true);
    }
  }, [isShow]);

  return (
    <TabPanelWrapper
      {...other}
      className={classNames('sinoui-tab-panel', className)}
    >
      {isRendered ? children : null}
    </TabPanelWrapper>
  );
}
