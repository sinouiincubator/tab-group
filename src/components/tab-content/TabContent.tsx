import React, { useMemo } from 'react';
import classNames from 'classnames';
import TabContentContext from '../TabContentContext';
import TabContentWrapper from './TabContentWrapper';
import TabPanelListWrapper from './TabPanelListWrapper';
import useTabList from '../commons/useTabList';
import TabListContext from '../TabListContext';

export interface Props {
  children?: React.ReactNode;
  className?: string;
  /**
   * 设置当前选中的标签页。这是一个从 `0` 开始的索引，第一个标签页的索引为 `0`，第二个标签页的索引为 `1`，……
   */
  selectedIndex: number;
  /**
   * 默认情况下只有当前标签面板会渲染到 DOM 中（出于性能考量）。如果设置为 `true`，则所有的标签面板从一开始就会立即渲染到 DOM 中。默认为 `false`。
   */
  forceRenderTabPanel?: boolean;
  /**
   * 设置启动标签面板渲染到 DOM 的缓存特性。当`forceRenderTabPanel` 为 `false` 此配置才有效。默认为 `true`。如果设置为 `false`，则不会缓存标签面板的渲染，标签切换后，销毁该标签面板的DOM。
   */
  cacheable?: boolean;
}

/**
 * 选项卡组件的内容组件
 */
export default function TabContent(props: Props) {
  const {
    className,
    selectedIndex,
    children,
    forceRenderTabPanel,
    cacheable,
    ...rest
  } = props;
  const tabListContext = useTabList(selectedIndex);
  const context = useMemo(
    () => ({ inTabContent: true, forceRenderTabPanel, cacheable }),
    [cacheable, forceRenderTabPanel],
  );
  const transform = useMemo(
    () => `translate3d(-${100 * selectedIndex}%, 0px, 0px)`,
    [selectedIndex],
  );

  return (
    <TabListContext.Provider value={tabListContext}>
      <TabContentContext.Provider value={context}>
        <TabContentWrapper
          className={classNames('sinoui-tab-content', className)}
          {...rest}
        >
          <TabPanelListWrapper
            style={{ transform }}
            className="sinoui-tab-panel-list"
          >
            {children}
          </TabPanelListWrapper>
        </TabContentWrapper>
      </TabContentContext.Provider>
    </TabListContext.Provider>
  );
}
