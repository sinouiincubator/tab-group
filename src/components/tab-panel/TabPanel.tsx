/* eslint-disable react/destructuring-assignment */
import React, { useContext, useRef, useEffect } from 'react';
import classNames from 'classnames';
import TabPanelWrapper from './TabPanelWrapper';
import TabListContext from '../TabListContext';
import useTabRegister from '../commons/useTabRegister';
import TabContentContext from '../TabContentContext';

interface TabPanelProps {
  children: React.ReactNode;
  className?: string;
  style?: React.CSSProperties;
  /**
   * 指定标签 id
   */
  tabId?: string;
  /**
   * 默认情况下，只有当前标签内容会渲染到 DOM 中。设置为 `true`，会在组件初始化时立即将标签内容渲染到 DOM 中。
   */
  forceRenderContent?: boolean;
  /**
   * 设置启用标签面板渲染到 DOM 的内容缓存。默认为 `true`。
   */
  cacheable?: boolean;
}

/**
 * 获取是否启用标签渲染缓存特性的hook。
 *
 * @param isCacheable 是否启用缓存特性的属性
 */
function useIsCacheable(isCacheable?: boolean) {
  const tabContentContext = useContext(TabContentContext);

  if (typeof isCacheable === 'boolean') {
    return isCacheable;
  }

  if (tabContentContext && typeof tabContentContext.cacheable === 'boolean') {
    return tabContentContext.cacheable;
  }

  return true;
}

/**
 * 判断是否需要渲染标签面板的hook。
 *
 * @param isCacheable 是否启用了标签面板内容缓存
 * @param isActive 是否是当前标签面板
 */
function useIsNeedRendered(isCacheable: boolean, isActive: boolean) {
  const isNeedRenderedRef = useRef(false);

  const isNeedRendered = isActive || (isCacheable && isNeedRenderedRef.current);

  useEffect(() => {
    if (isActive && isCacheable) {
      isNeedRenderedRef.current = true;
    }
  });

  return isNeedRendered;
}

/**
 * 判断是否是当前面板面板的hook。
 *
 * @param index 标签面板索引位置
 */
function useIsActive(index: number) {
  const context = useContext(TabListContext);
  const selectedIndex = context ? context.selectedIndex : 0;

  return index === selectedIndex;
}

/**
 * 判断是否强制渲染标签面板内容的hook。
 *
 * @param isForceRenderContent 是否强制渲染的属性
 */
function useIsForceRenderContent(isForceRenderContent?: boolean) {
  const tabContentContext = useContext(TabContentContext);

  if (typeof isForceRenderContent === 'boolean') {
    return isForceRenderContent;
  }
  return !!tabContentContext && tabContentContext.forceRenderTabPanel;
}

/**
 * 标签页面板组件
 */
export default function TabPanel(props: TabPanelProps) {
  const {
    children,
    className,
    forceRenderContent,
    cacheable,
    ...other
  } = props;
  const index = useTabRegister();
  const isActive = useIsActive(index);
  const isCacheable = useIsCacheable(cacheable);
  const isNeedRendered = useIsNeedRendered(isCacheable, isActive);
  const isForceRenderContent = useIsForceRenderContent(forceRenderContent);

  if (index === -1) {
    return null;
  }

  return (
    <TabPanelWrapper
      data-testid={`tab-panel-${index}`}
      role="tabpanel"
      aria-hidden={!isActive}
      aria-labelledby={props.tabId}
      {...other}
      className={classNames(
        'sinoui-tab-panel',
        {
          'sinoui-tab-panel-active': isActive,
          'sinoui-tab-panel-hidden': !isActive,
        },
        className,
      )}
    >
      {isNeedRendered || isForceRenderContent ? children : null}
    </TabPanelWrapper>
  );
}
