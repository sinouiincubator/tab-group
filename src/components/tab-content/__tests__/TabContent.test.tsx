import { render, cleanup } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { act } from 'react-dom/test-utils';
import TabContent from '../TabContent';
import TabPanel from '../../tab-panel';

afterEach(cleanup);

it('渲染TabContent', () => {
  const { container } = render(<TabContent selectedIndex={1}>内容</TabContent>);

  const tabContent = container.querySelector(
    '.sinoui-tab-panel-list',
  ) as HTMLElement;

  expect(tabContent).toBeInTheDOM();
  expect(tabContent).toHaveStyle('transform: translate3d(-100%, 0px, 0px);');
});

/* 镜像测试 */

describe('TabContent 镜像测试', () => {
  it('渲染 TabContent', () => {
    const tree = renderer
      .create(<TabContent selectedIndex={1}>内容</TabContent>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

/* 验收测试 */
describe('TabContent + TabPanel 组合使用', () => {
  it('渲染 TabContent', () => {
    const { getByTestId } = render(
      <TabContent selectedIndex={1}>
        <TabPanel>内容1</TabPanel>
        <TabPanel>内容2</TabPanel>
        <TabPanel>内容3</TabPanel>
      </TabContent>,
    );

    expect(getByTestId('tab-panel-1')).toHaveClass('sinoui-tab-panel-active');
  });

  it('更换选中的标签页', () => {
    const { getByTestId, rerender } = render(
      <TabContent selectedIndex={1}>
        <TabPanel>内容1</TabPanel>
        <TabPanel>内容2</TabPanel>
        <TabPanel>内容3</TabPanel>
      </TabContent>,
    );

    act(() => {
      rerender(
        <TabContent selectedIndex={2}>
          <TabPanel>内容1</TabPanel>
          <TabPanel>内容2</TabPanel>
          <TabPanel>内容3</TabPanel>
        </TabContent>,
      );
    });

    expect(getByTestId('tab-panel-2')).toHaveClass('sinoui-tab-panel-active');
  });
});
