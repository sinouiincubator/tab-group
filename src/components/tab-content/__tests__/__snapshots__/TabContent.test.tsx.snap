// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabContent 镜像测试 渲染 TabContent 1`] = `
.c0 {
  overflow-x: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-transition: -webkit-transform 0.35s cubic-bezier(0.15,0.3,0.25,1) 0s;
  -webkit-transition: transform 0.35s cubic-bezier(0.15,0.3,0.25,1) 0s;
  transition: transform 0.35s cubic-bezier(0.15,0.3,0.25,1) 0s;
  will-change: transform;
}

.c1 > .sinoui-tab-panel-hidden {
  height: 0px;
}

<div
  className="c0 sinoui-tab-content"
>
  <div
    className="c1 sinoui-tab-panel-list"
    style={
      Object {
        "transform": "translate3d(-100%, 0px, 0px)",
      }
    }
  >
    内容
  </div>
</div>
`;
